features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/voicemessages/timeline/VoiceMessagePresenter.kt

```diff
@@ -0,0 +1,152 @@
+        override fun create(content: TimelineItemVoiceContent): VoiceMessagePresenter
+    }
+
+    private val voiceCache = voiceMessageCacheFactory.create(mxcUri = content.mediaSource.url)
+
+    private val player = voiceMessagePlayerFactory.create(
+        eventId = content.eventId,
+        mediaPath = voiceCache.cachePath
+    )
+
+    @Composable
+    override fun present(): VoiceMessageState {
+
+        val scope = rememberCoroutineScope()
+
+        val playerState by player.state.collectAsState(VoiceMessagePlayer.State(isPlaying = false, isMyMedia = false, currentPosition = 0L))
+        val mediaFile = remember { mutableStateOf<Async<MediaFile>>(Async.Uninitialized) }
+
+        val button by remember {
+            derivedStateOf {
+                when {
+                    content.eventId == null -> VoiceMessageState.Button.Disabled
+                    playerState.isPlaying -> VoiceMessageState.Button.Pause
+                    mediaFile.value is Async.Loading -> VoiceMessageState.Button.Downloading
+                    mediaFile.value is Async.Failure -> VoiceMessageState.Button.Retry
+                    else -> VoiceMessageState.Button.Play
+                }
+            }
+        }
+        val progress by remember {
+            derivedStateOf { if (playerState.isMyMedia) playerState.currentPosition / content.duration.toMillis().toFloat() else 0f }
+        }
+        val time by remember {
+            derivedStateOf {
+                val time = if (playerState.isMyMedia) playerState.currentPosition else content.duration.toMillis()
+                "%02d:%02d".format(time.msToMins(), time.msToSecs())
+            }
+        }
+
+        suspend fun downloadCacheAndPlay() {
+            mediaFile.runUpdatingState {
+                mediaLoader.downloadMediaFile(
+                    source = content.mediaSource,
+                    mimeType = content.mimeType,
+                    body = content.body,
+                ).mapCatching {
+                    if (voiceCache.moveToCache(it.toFile())) {
+                        player.acquireControlAndPlay()
+                        it
+                    } else {
+                        error("Failed to move file to cache.")
+                    }
+                }
+            }
+        }
+
+        fun eventSink(event: VoiceMessageEvents) {
+            when (event) {
+                is VoiceMessageEvents.PlayPause -> {
+                    if (playerState.isMyMedia) {
+                        if (playerState.isPlaying) {
+                            player.pause()
+                        } else {
+                            player.play()
+                        }
+                    } else {
+                        if (voiceCache.isInCache()) {
+                            player.acquireControlAndPlay()
+                        } else {
+                            scope.launch { downloadCacheAndPlay() }
+                        }
+                    }
```

Comment:
The ownership aspect is a lot for the presenter to know about. Could the `VoiceMessagePlayer` not take care of some of this logic? Leaving us with something simpler like

```kotlin
if (!isDownloaded) {
   download()
}

if (isPlaying) {
   pause()
} else {
   play()
}
```

Reply:
I agree, but I prefer to take some more time aside to make this better.

Reply:
Of course, I don't think it should block this PR

Reply:
Of course, I don't think it should block this PR